destination/files.go

```diff
@@ -12,45 +14,37 @@ import (
 	"github.com/klauspost/compress/zstd"
 )
 
-type ReadCSVFileResultType int
-
-const (
-	Success ReadCSVFileResultType = iota
-	KeyNotFound
-	FileNotFound
-	FailedToDecompress
-	FailedToDecrypt
-)
-
-type ReadCSVFileResult struct {
-	Type  ReadCSVFileResultType
-	Data  *[]byte // only set if Type is Success
-	Error *error  // only set if Type is FailedToDecrypt or FailedToDecompress
-}
-
 func ReadCSVFile(
 	fileName string,
 	keys map[string][]byte,
 	compression pb.Compression,
 	encryption pb.Encryption,
-) *ReadCSVFileResult {
+) (CSV, error) {
 	key, ok := keys[fileName]
 	if !ok {
-		return &ReadCSVFileResult{Type: KeyNotFound}
+		return nil, fmt.Errorf("key for file %s not found", fileName)
 	}
 	file, err := os.ReadFile(fileName)
 	if err != nil {
-		return &ReadCSVFileResult{Type: FileNotFound}
+		return nil, fmt.Errorf("file %s does not exist", fileName)
 	}
 	decrypted, err := Decrypt(key, file, encryption)
 	if err != nil {
-		return &ReadCSVFileResult{Type: FailedToDecrypt, Error: &err}
+		return nil, fmt.Errorf("failed to decrypt file %s, cause: %w", fileName, err)
 	}
 	decompressed, err := Decompress(decrypted, compression)
 	if err != nil {
-		return &ReadCSVFileResult{Type: FailedToDecompress, Error: &err}
+		return nil, fmt.Errorf("failed to decompress file %s, cause: %w", fileName, err)
 	}
```

Comment:
it's worth to stream the data using `io.Reader` interface

Reply:
Let's do it as a follow-up (#5)